spring-kafka-docs/src/main/kotlin/org/springframework/kafka/kdocs/requestreply/Application.kt

```diff
@@ -0,0 +1,136 @@
+import org.apache.kafka.common.serialization.ByteArrayDeserializer
+import org.apache.kafka.common.serialization.ByteArraySerializer
+import org.slf4j.LoggerFactory
+import org.springframework.boot.SpringApplication
+import org.springframework.context.annotation.Bean
+import org.springframework.kafka.core.ProducerFactory
+import java.util.concurrent.TimeUnit
+import java.util.function.Consumer
+
+/**
+ * Code snippets for request/reply messaging.
+ *
+ * @author Gary Russell
+ * @since 2.7
+ */
+@SpringBootApplication
+class Application {
+    @Bean
+    fun topics(): NewTopics {
+        return NewTopics(
+            TopicBuilder.name("requests")
+                .partitions(10)
+                .replicas(1)
+                .build(),
+            TopicBuilder.name("replies")
+                .partitions(10)
+                .replicas(1)
+                .build()
+        )
+    }
+
+    /* This is ok in Intellij but fails in gradle
+    @Bean
+    fun kafkaTemplate(pf: ProducerFactory<Any, Any>?): KafkaTemplate<Any, Any> {
+        return KafkaTemplate(pf)
+    }
+     */
+
+// tag::beans[]
+    @Bean
+    fun template(
+        pf: ProducerFactory<String?, String>?,
+        factory: ConcurrentKafkaListenerContainerFactory<String?, String?>
+    ): ReplyingKafkaTemplate<String?, String, String?> {
+        val replyContainer = factory.createContainer("replies")
+        replyContainer.containerProperties.groupId = "request.replies"
+        val template = ReplyingKafkaTemplate(pf, replyContainer)
+        template.messageConverter = ByteArrayJsonMessageConverter()
+        template.defaultTopic = "requests"
+        return template
+    }
+// end::beans[]
+
+    @Bean
+    fun runner(template: ReplyingKafkaTemplate<String?, String?, String?>): ApplicationRunner {
+        return ApplicationRunner { _ ->
+// tag::sendReceive[]
+            val future1: RequestReplyTypedMessageFuture<String?, String?, Thing?>? =
+                template.sendAndReceive(MessageBuilder.withPayload("getAThing").build(),
+                    object : ParameterizedTypeReference<Thing?>() {})
+            log.info(future1?.sendFuture?.get(10, TimeUnit.SECONDS)?.recordMetadata?.toString())
+            val thing = future1?.get(10, TimeUnit.SECONDS)?.payload
+            log.info(thing.toString())
+
+            val future2: RequestReplyTypedMessageFuture<String?, String?, List<Thing?>?>? =
+                template.sendAndReceive(MessageBuilder.withPayload("getThings").build(),
+                    object : ParameterizedTypeReference<List<Thing?>?>() {})
+            log.info(future2?.sendFuture?.get(10, TimeUnit.SECONDS)?.recordMetadata.toString())
+            val things = future2?.get(10, TimeUnit.SECONDS)?.payload
+            things?.forEach(Consumer { thing1: Thing? -> log.info(thing1.toString()) })
+// end::sendReceive[]
+        }
+    }
+
+    @KafkaListener(id = "myId", topics = ["requests"], properties = ["auto.offset.reset:earliest"])
```

Comment:
So weird that Kotlin cannot determine a single value for annotation array attribute 😄 
Probably some interoperability with Java...